src/java/org/apache/cassandra/index/Index.java

```diff
@@ -657,36 +664,36 @@ default PartitionIterator filterReplicaFilteringProtection(PartitionIterator ful
      * {@link SecondaryIndexManager#registerIndex(Index, Object, Supplier)} during index registering
      * at {@link #register(IndexRegistry)} method.
      */
-    interface Group
+    interface Group<T extends Index>
```

Comment:
So this definitely seemed like a good idea initially, but after looking at everything else it touches outside of SAI, we might want to reconsider. With a type argument on `Group`, we end up w/ tons of non-trivial compile warnings to resolve in `SecondaryIndexManager` (ex. `registerIndex()`), and with its methods like `listIndexGroups()` returning raw-typed groups to clients, you have raw type warnings in `CompactionManager`, `SSTableWriter`, etc.

Little assertions in `removeIndex()` and `containsIndex()` are probably the lesser evil. Sorry for my myopia...

Reply:
It's a pity but I agree. I have addressed the initial issue by changing the index set in `StorageAttachedIndexGroup` to `Set<Index> indices`. This avoids the need to copy the set in `getIndexes`.